pkg/operators/filter/filter.go

```diff
@@ -0,0 +1,430 @@
+        `,
+		Alias: "F",
+	}}
+}
+
+func (f *filterOperator) InstantiateDataOperator(gadgetCtx operators.GadgetContext, instanceParamValues api.ParamValues) (operators.DataOperatorInstance, error) {
+	filterCfg, _ := instanceParamValues[ParamFilter]
+
+	gadgetCtx.Logger().Debugf("logger %+v", instanceParamValues)
+
+	fop := &filterOperatorInstance{
+		ffns: map[datasource.DataSource][]func(datasource.DataSource, datasource.Data) bool{},
+	}
+
+	filters := strings.Split(filterCfg, ",")
+	for _, filter := range filters {
+		if filter == "" {
+			continue
+		}
+		gadgetCtx.Logger().Debugf("adding filter %q", filter)
+		err := fop.addFilter(gadgetCtx, filter)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return fop, nil
+}
+
+func (f *filterOperator) Priority() int {
+	return Priority
+}
+
+type filterOperatorInstance struct {
+	ffns map[datasource.DataSource][]func(datasource.DataSource, datasource.Data) bool
+}
+
+func (f *filterOperatorInstance) Name() string {
+	return name
+}
+
+func (f *filterOperatorInstance) PreStart(gadgetCtx operators.GadgetContext) error {
+	for ds, funcs := range f.ffns {
+		funcs := funcs
+		ds.Subscribe(func(ds datasource.DataSource, data datasource.Data) error {
+			for _, fn := range funcs {
+				if !fn(ds, data) {
+					return datasource.ErrDiscard
+				}
+			}
+			return nil
+		}, Priority) // TODO: need some predefined & sane values
+	}
+	return nil
+}
+
+func (f *filterOperatorInstance) Start(gadgetCtx operators.GadgetContext) error {
+	return nil
+}
+
+func (f *filterOperatorInstance) Stop(gadgetCtx operators.GadgetContext) error {
+	return nil
+}
+
+func getCompareFunc[T constraints.Ordered](op comparisonType) func(a, b T) bool {
+	switch op {
+	default:
+		return func(a, b T) bool {
+			return false
+		}
+	case comparisonTypeMatch:
+		return func(a, b T) bool {
+			return a == b
+		}
+	case comparisonTypeLt:
+		return func(a, b T) bool {
+			return a < b
+		}
+	case comparisonTypeGt:
+		return func(a, b T) bool {
+			return a > b
+		}
+	case comparisonTypeLte:
+		return func(a, b T) bool {
+			return a <= b
+		}
+	case comparisonTypeGte:
+		return func(a, b T) bool {
+			return a >= b
+		}
+	}
+}
+
+func extractFilter(filter string) (dsName string, fieldName string, op comparisonType, negate bool, value string, err error) {
+	// State machine to get filter
+	var opString string
+
+	stage := 0
+	pos := 0
+nextChar:
+	for {
+		if pos >= len(filter) {
+			break
+		}
+		switch stage {
+		case 0:
+			switch filter[pos] {
+			case ':':
+				dsName = fieldName
+				fieldName = ""
+				pos++
+				continue nextChar
+			case '!', '~', '>', '<', '=':
+				stage = 1
+				continue nextChar
+			}
+			fieldName += string(filter[pos])
+			pos++
+		case 1:
+			switch filter[pos] {
+			case '!', '~', '>', '<', '=':
+				opString += string(filter[pos])
+				pos++
+			default:
+				switch opString {
+				case "=", "==":
+					op = comparisonTypeMatch
+				case "!=":
+					op = comparisonTypeMatch
+					negate = true
```

Comment:
Perhaps we can have two new `comparisonTypeNotMatch` and `comparisonTypeNonMatchRegex` and remove `negate altogether?

Reply:
@burak-ok suggested the same - I think if we would add a programmatic approach later on, it makes sense to just negate a full "rule" instead of having a couple of explicit negations that can't be use for everything.
I don't have a super strong opinion about it, but to me this feels more complete / future proof.

Reply:
ok, let's keep the negate if you feel it'll help later on.